ui/radar/radar.js

```diff
@@ -0,0 +1,251 @@
+      options = Object.assign({}, this.options, options.RankOptions[monster['rank']]);
+    let tr = monster['dom'];
+    // calculate rotation based on facing
+    let playerVector = new Point2D(e.detail.pos.x, -e.detail.pos.y);
+    let targetVector = monster['pos'];
+    let deltaVector = new Point2D(targetVector.x - playerVector.x, targetVector.y - playerVector.y);
+    if (tr) {
+      tr.childNodes[1].innerHTML = monster['rank'] + '&nbsp;&nbsp;&nbsp;&nbsp;' + monster['name'];
+      if (Math.abs(e.detail.pos.z - monster['pos_z']) > 5)
+        tr.childNodes[1].innerHTML += '&nbsp;&nbsp;' + (e.detail.pos.z < monster['pos_z']? '↑' : '↓');
+      tr.childNodes[1].innerHTML += '<br>' + length(deltaVector).toFixed(2) + 'm';
+      if (Date.now() / 1000 <= monster['battle_time'] + 60) {
+        tr.childNodes[1].innerHTML += ' ' + (monster['current_hp'] * 100 /
+          monster['hp']).toFixed(2) + '%';
+      }
+      if (options.Puller) {
+        if (monster['puller'])
+          tr.childNodes[1].innerHTML += '&nbsp;&nbsp;' + monster['puller'];
+      }
+      if (options.Position) {
+        tr.childNodes[1].innerHTML += '<br>X: ' +
+          posToMap(parseFloat(monster['pos_x'])).toFixed(1) + '&nbsp;&nbsp;Y:' +
+          posToMap(parseFloat(monster['pos_y'])).toFixed(1);
+      }
+    }
+    if (options.DetectionRange > 0 && length(deltaVector) > options.DetectionRange)
+      monster['dom'].setAttribute('class', 'hide');
+    else
+      monster['dom'].setAttribute('class', '');
+    let deltaTheta = Math.acos(theta(deltaVector, new Point2D(1, 0)));
+    if (deltaVector.y < 0)
+      deltaTheta = -deltaTheta;
+    deltaTheta += Math.PI - e.detail.rotation;
+    let angle = -deltaTheta*180/Math.PI;
+    let arrow_id = 'arrow-' + monster['id'];
+    let arrow = document.getElementById(arrow_id);
+    arrow.style.transform='rotate('+angle+'deg)';
+  }
+
+  RemoveMonster(mob_name) {
+    if (mob_name in this.targetMonsters) {
+      this.targetMonsters[mob_name]['dom'].remove();
+      delete this.targetMonsters[mob_name];
+    }
+  }
+
+  OnLogEvent(e) {
+    let lang = this.options.Language;
+    for (let i = 0; i < e.detail.logs.length; i++) {
+      // add new combatant
+      let matches = e.detail.logs[i].match(Regexes.addedCombatantFull());
+      if (matches) {
+        let monster = this.nameToMonster[matches.groups.name];
+        if (monster)
+          this.AddMonster(e.detail.logs[i], monster, matches);
+      }
+      // network ability
+      matches = e.detail.logs[i].match(Regexes.abilityFull({}));
+      if (matches) {
+        let monster = this.targetMonsters[matches.groups.target];
+        if (monster)
+          this.UpdateMonsterPuller(monster, matches.groups.source);
+      }
+      // change instances
+      let r = e.detail.logs[i].match(instanceChangedRegex[lang] || instanceChangedRegex['en']);
+      if (r)
+        this.ClearTargetMonsters(10); // don't remove mobs lasting less than 10 seconds
```

Comment:
Does this mean they will stick around forever?

Reply:
At first, it will clear all monsters when changeZone or changeInstance happens. And  then there's a bug that because changeInstance is implemented by matching the `00` log by `instanceChangedRegex `, and such log will appear after several seconds when we enter a map. 

So, if radar detects monsters before the log shows up (which means the monster is just near the teleportation crystal or it's an S rank), it will clear the detected monster by the entering instance log.